#286: Fixed distributed writing to local filesystem

fhueske opened the pull request:
This fix enables local filesystem writes in a cluster setup.
Output directories for local FS writes are created on each TM but not on the JM (except if it hosts a TM as well).

The currently implemented override policy for local filesystem writes is:
- existing files is not deleted to create output directories (to avoid deleting OS files)
- existing files in existing output directories are overwritten

Let me know, if you want to enable deleting local FS files to create output directories.

supermegaciaccount committed:
implemented jdbcoutputformat

Commit:
Introduced CREATE and OVERWRITE file output write modes. OVERWRITE is default (as before).
Rewrote code for file output path preparation.
Fixed distributed writing to local filesystem #286

rmetzger commented on stratosphere-runtime/src/main/java/eu/stratosphere/pact/runtime/task/DataSinkTask.java:
```diff
@@ -385,43 +385,52 @@ public int getMaximumNumberOfSubtasks()
-				if (f.isDir()) {
-					return -1;
+			
+			if(fs.getUri().getScheme().toLowerCase().equals("hdfs") || fs.getUri().getScheme().toLowerCase().equals("s3")) {
```
I'm not sure if this is the correct place for this.
I thought that file systems are pluggable. This here looks like hardcoded rules for specific file systems.

is there a way to move this code into the DistributedFileSystem (and S3) implementation?

The `FileOutputFormat` also contains specific `file:///` scheme code. I'm think In/Outputformats are independent of the filesystems (e.g. the file systems provide a data stream and the IF/OF converts it into an internal representation?)

fhueske replied:
You are certainly right, that this is not the optimal place. In fact, most of that was already present and I only added a guard for distributed FS at this place. I was looking for a quick fix that doesn't break anything shortly before the release and didn't want to change too many things.

Now with the release out of the way, I can look for a proper solution based on file systems. However, the distinction between distributed and local file system will most likely remain in the DataSinkTask and the FileInputFormat as the distributed FS should be initialized at planning time (within the JM) and local FS need to be initialized in parallel at runtime...

rmetzger replied:
Okay. I also think that it is a better idea to implement it correctly (e.g. no quick fix).

For the initialization at planning time: Can't you just add a new method to the FileSystems class for that?

markus-h committed:
Polling for jobmanager webinterface

markus-h committed:
Polling for jobmanager webinterface

Commit:
Changed version to 0.4

Commit:
Changed version to 0.5-SNAPSHOT

Commit:
Applied version changes to all pom files.

Commit:
Renamed 'match' in functions to 'join'.

Commit:
Fixed bug in web-client.

supermegaciaccount committed:
moved jdbc into new project, adjusted for 0.5

Conflicts:
	stratosphere-addons/jdbc/pom.xml
	stratosphere-addons/jdbc/src/main/java/eu/stratosphere/api/io/jdbc/JDBCInputFormat.java
	stratosphere-addons/jdbc/src/test/java/eu/stratosphere/api/io/jdbc/JDBCInputFormatTest.java
Conflicts handled by @rmetzger

Commit:
Rename FilterStub to FilterFunction

rmetzger committed:
Set LocalExecutor logging to INFO; print runtime. reduce logging verbosity of IPC server

Commit:
Clean up bytes to memory pages computations. Fix number overflow errors for very large memory sizes.

Commit:
Clean up destriptions and log/web display names.

markus-h committed:
Paths in windows scripts enclosed in double quotes

rmetzger committed:
Created 'avro' addons subproject
Added 'AvroInputFormat' with test.

Commit:
Add shutdown method to LocalDistributedExecutor and remove failing CustomDataTypeTest
=> the removed test functionality will be included in an upcoming test by @aljoscha

twalthr committed:
Avro complex types (array, map, enum) implemented

rmetzger committed:
Undo eclipse formatting in avro if
disable javadocs gen when java6 is being used.

aljoscha committed:
Fix bug in scala cogroup.flatMap

Did fail with null pointer exception when one of the inputs is empty.

rmetzger committed:
renamed jdbc package properly to the new 0.4 naming conventions

rmetzger committed:
added batch interval to jdbc; fixed issue with jdbc test cases

Commit:
Merge branch 'webinterface_autoupdate_newrelease-pr' of https://github.com/markus-h/stratosphere into webfix

Commit:
Fix bug that incorrectly marked iterative tasks.

Commit:
Added standalone plan visualizer tool.

Commit:
Add LineRank as test case.

Commit:
Add Scala Page Rank examples

Commit:
Added test for write modes on local file system

fhueske commented:
New PR for this: #421
